src/freelunch/darwin.py

```diff
@@ -167,35 +104,73 @@ def op(self, *parents, pop=None, F=hypers['F']):
         return x.dna + F * (best.dna - a.dna) + F * (b.dna - c.dna)
 
 
-# Exportable dictionary
-DE_methods = {x.name: x for x in [
-    DE_rand_1, DE_rand_2, DE_best_1, DE_best_2, DE_current_to_best_1]}
-
-
-# API for probability update
-def update_strategy_ps(strats):
-    hits = np.array([s.hits[-1] for s in strats])
-    wins = np.array([s.wins[-1] for s in strats])
-    total_hits = np.sum(hits)
-    total_wins = np.sum(wins)
-    ps = np.zeros_like(hits)
-    # update model
-    # prevent zero division with a bit of fun bias!!
-    dem = np.sum(wins * (hits + wins)) + 1
-    for i, h, w in zip(range(len(ps)), hits, wins):
-        num = h * (total_hits + total_wins)
-        ps[i] = num / dem
-    # normalise
-    n = np.sum(ps)
-    if n == 0:
-        ps += 1
-    ps = ps / n
-    for strat, p in zip(strats, ps):
-        strat.update(p)
-
-
-# API for strategy selection 
-def select_strategy(strats):
-    ps = [s.p[-1] for s in strats]
-    ps = ps/np.sum(ps)
-    return np.random.choice(strats, p=ps)
\ No newline at end of file
+
+# %% Crossover operations
+
+class Crossover():
+    ''' 
+    Generic crossover operation
+    '''
+
+    name='Crossover'
+    hypers={}
+
+    def __init__(self,hypers=hypers):
+        self.hypers = hypers
+
+    def breed(self, parent1, parent2):
+        raise NotImplementedError
+
+class XOver_binary(Crossover):
+
+    name='Binary Crossover'
+    hypers={
+        'Cr':0.2
+    }
+
+    def breed(self, parent1, parent2, p=hypers['Cr']):
+        out = np.empty_like(parent1)
+        for a, b, i in zip(parent1, parent2, range(len(parent1))):
+            if np.random.uniform(0, 1) < p:
+                out[i] = a
+            else:
+                out[i] = b
+        #Ensure at least one difference
+        jrand = np.random.randint(0, len(out))
+        out[jrand] = parent2[jrand]
+        return out
+
+
+
+# %% Module Functions
+
+
+# All DE mutation methods defined prior to this
+def parse_adaptable_search(op):
+    '''
+    Parsing different search operations as strings or handles
+    '''
+    if isinstance(op, adaptable_search_operation):
+        return op
+    elif isinstance(op, str):
+        op = op.replace('/','_')
+        if op.startswith('DE_'):
+            return globals()[op]
```

Comment:
I'm definitely reticent to use globals like this... I would prefer to specifically hardcode a list of functions/ops within the module and go for:

if op in permitted_ops:
    
    ---recover the op---
    
    return op

Reply:
The reason is that globals[op] can return anything in globals and not just the methods we want...

Reply:
Oh I see we use the DE_ handle to filter... yeah I guess that works fine